### PR TITLE
CrowdStrike Malquery Integration

### PR DESCRIPTION
**Pull request by Bargenish**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: demisto/etc/issues/23499

## Description
Use MalQuery Integration to query the contents of over a half-billion binary files, both clean and malicious, that are part of Falcon MalQuery's corpus.
Done according to the new documentation using Oauth2 and not the one that was attached in the issue

## Minimum version of Demisto
- [ ] 4.5.0
- [X] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [X] Tests
- [X] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.yml [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        type: String
+      - contextPath: File.MD5
+        description: MD5 hash of the file.
+        type: String
+  - name: file
+    arguments:
+    - name: file
+      required: true
+      description: The file SHA256.
+      isArray: false
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.yml:311]: should support multiple values

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.yml [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+      isArray: true
+    outputs:
+    - contextPath: Malquery.Request_ID
+      description: The request ID.
+      type: String
+    description: "Use hunt to schedule a YARA-based search for execution. Results\
+      \ can be filtered based on criteria such as file type, file size and first seen\
+      \ date.     Using hunt is a two-step process:     \0Schedule a YARA-based search\
+      \ with hunt, which returns a request ID. Provide that request ID to the cs-malquery-get-request\
+      \ to retrieve results."
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.yml:107]: spacing looks weird, make sure it looks ok in the platform

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+    try:
+        content = raw_response.content
+    except Exception as e:
+        raise DemistoException(
+            f"Failed to load file data. \n {e}")
+
+    return fileResult(file_id, content)
+
+
+# Malquery counts the download as the number of sha256 passed to the endpoind and not as a single download.
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:338]:
```suggestion
# Malquery counts the download as the number of sha256 passed to the endpoint and not as a single download.
```

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+    return CommandResults(
+        readable_output=human_readable,
+        outputs_prefix='Malquery',
+        outputs_key_field='Request_ID',
+        outputs=entry_context,
+        raw_response=raw_response)
+
+
+def get_file_metadata_command(client: Client, args: dict):
+    indicator_type = 'File'
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:302]: why do you need this variable? it's being used once in a string formatting

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+        entry_context = {
+            "Request_ID": request_id,
+            "Status": status
+        }
+        human_readable = tableToMarkdown('Request Status:', entry_context, removeNull=True)
+    else:
+        entry_context = {
+            "Request_ID": request_id,
+            "Status": status,
+            "File": resources if len(resources) else None
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:289]:
```suggestion
            "File": resources if resources else None
```

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+            "type": key,
+            "value": args[key]
+        } for key in pattern_names if args.get(key)
+    ]
+    # must provide a pattern (hex, ascii ot wide string)
+    if not patterns:
+        raise DemistoException("You must provide a query to search in the following patterns: Hex, ASCII, Wide string")
+
+    query_filters = {
+        "limit": int(args.get('limit', '')) if args.get('limit') else None,
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:229]: you can decide on a different default value.
is there an option for pagination?
```suggestion
        "limit": int(args.get('limit', '100')),
```

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+            return access_token
+
+    def get_token_request(self):
+        """
+            Sends token request
+
+            :rtype ``str``
+            :return: Access token
+        """
+        # body = f"client_id={self.client_id}&client_secret={self.client_secret}"
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:107]: remove

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+           Obtains access and refresh token from server.
+           Access token is used and stored in the integration context until expiration time.
+           After expiration, new refresh token and access token are obtained and stored in the
+           integration context.
+
+           Returns:
+               str: Access token that will be added to authorization header.
+       """
+        now = datetime.now()
+        integration_context = demisto.getIntegrationContext()[0] \
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:81]: why would it be a list?

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+        Returns the time passed in mins
+        :param start_time: Start time in datetime
+        :param end_time_str: End time in str
+        :return: The passed mins in int
+    """
+    time_delta = start_time - datetime.fromtimestamp(end_time_str, tz)
+    return time_delta.seconds / 60
+
+
+def remove_None_values_keys(d):
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:34]: use `assign_params` from `CommonServerPython` instead

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py [hunk 10]
```diff
@@ -0,0 +1,425 @@
+import demistomock as demisto
+from CommonServerPython import *  # noqa: E402 lgtm [py/polluting-import]
+from CommonServerUserPython import *  # noqa: E402 lgtm [py/polluting-import]
+
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+# CONSTANTS
+DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+VENDOR_NAME = 'CrowdStrike Malquery'
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:10]: should be exactly as the integration ID (or add the space to the id)
```suggestion
VENDOR_NAME = 'CrowdStrikeMalquery'
```

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.yml (hunk 1 above)
Comment (Bargenish) [reply Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.yml:311]: @yaakovi - talked about it with @Arsenikr. He told me that although the API allows it, the command should only support 1 file at a time.

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py (hunk 6 above)
Comment (Bargenish) [reply Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:229]: @yaakovi  I looked at the response model in the swagger file and there is an option for pagination but on the actual response there wasn't a `pagination` field as supposed and it also didn't appear on the malquery documentation.

File: Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py (hunk 8 above)
Comment (Bargenish) [reply Packs/CrowdStrikeMalquery/Integrations/CrowdStrikeMalquery/CrowdStrikeMalquery.py:81]: @yaakovi - when I tested the integration using the demisto plugin on pycharm, it failed because the integration context was a list so I added the option.
should I remove it?

File: package-lock.json [hunk 11]
```diff
@@ -687,15 +687,14 @@
       "integrity": "sha512-a30VEBm4PEdx1dRB7MFK7BejejvCvBronbLjht+sHuGYj8PHs7M/5Z+rt5lw551vZ7yfTCj4Vuyy3mSJytDWRQ=="
     },
     "handlebars": {
-      "version": "4.7.6",
-      "resolved": "https://registry.npmjs.org/handlebars/-/handlebars-4.7.6.tgz",
-      "integrity": "sha512-1f2BACcBfiwAfStCKZNrUCgqNZkGsAT7UM3kkYtXuLo0KnaVfjKOyf7PRzB6++aK9STyT1Pd2ZCPe3EGOXleXA==",
+      "version": "4.7.3",
```
Comment (yaakovi) [inline package-lock.json]: I meant to revert the changes to the file (all of them).
restore the entire file.

**Comment by yaakovi** [comment]
@yaron-libman @kirbles19 
please review integration and playbook for next content release (4.8)
@Bargenish test failed on 5.0

**Comment by Bargenish** [comment]
@yaakovi 👍  , I increased the timeout for the `fuzzy_search` request.

> @yaron-libman @kirbles19
> please review integration and playbook for next content release (4.8)
> @Bargenish test failed on 5.0

**Comment by yaron-libman** [comment]
@richardbluestone I reviewed the playbook yamls. When you're done with the integration, please add the docs-approved label.

**Comment by yaron-libman** [comment]
@Bargenish out of curiosity, are spaces allowed - this is the only place i can recall seeing a space in an input?

inputs.search command

File: Packs/CrowdStrikeMalquery/Playbooks/CrowdStrikeMalquery_-_GenericPolling_-_Search.yml [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+      "dimensions": {
+        "height": 910,
+        "width": 810,
+        "x": 50,
+        "y": 50
+      }
+    }
+  }
+inputs:
+- key: search command
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Playbooks/CrowdStrikeMalquery_-_GenericPolling_-_Search.yml:329]: @yaron-libman good catch, thanks.
it will work but better to keep up with the standard.

@Bargenish please rename it (use `snake_case`)

**Comment by Bargenish** [comment]
@yaron-libman There is no command called `inputs.search`. It is an input for the generic polling playbook. The user should use this as a sub playbook and pass the desired search command (`hunt` or `exact-search`).  
You can see an example in the test playbook, tasks 31 and 32.

File: Packs/CrowdStrikeMalquery/Playbooks/CrowdStrikeMalquery_-_GenericPolling_-_Search.yml (hunk 12 above)
Comment (Bargenish) [reply Packs/CrowdStrikeMalquery/Playbooks/CrowdStrikeMalquery_-_GenericPolling_-_Search.yml:329]: Done.
@yaron-libman thanks!

**Comment by richardbluestone** [comment]
Reviewed the yaml and md files

File: Packs/CrowdStrikeMalquery/Playbooks/CrowdStrikeMalquery_-_GenericPolling_-_Multidownload_and_Fetch.yml [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    task:
+      id: 31ae8182-92f3-45a5-865a-d691cdcdd275
+      version: -1
+      name: Samples multidownload
+      description: 'Schedule samples for download. Using samples-multidownload is
+        a three-step process: 1. Schedule the download with samples-multidownload,
+        which returns a request ID. 2. Provide that request ID to the cs-malquery-get-request,
+        in order to check the status of the operation. 3. When the request status
+        is “done”, use cs-malquery-sample-fetch to download the results as a password-protected
+        archive. The password to extract results from the archive is: 'infected''    '
```
Comment (yaakovi) [inline Packs/CrowdStrikeMalquery/Playbooks/CrowdStrikeMalquery_-_GenericPolling_-_Multidownload_and_Fetch.yml:61]:
```suggestion
        archive. The password to extract results from the archive is: "infected"'
```
